### DISCUSSION TITLE
Possible bug: Self-hosted Renovate does not work with NPM deps unless Node version is specified

### DISCUSSION THREAD

**Discussion by mikaello** [renovatebot/renovate#17874] [Suggest an Idea]
## How are you running renovate

Self-hosted ([whitesource/renovate-on-prem](https://github.com/whitesource/renovate-on-prem))

Runtime environment does not have internet access, everything has to go through Artifactory proxy (i.e. custom registries).

## What version of Renovate?

Version [Whitesource Renovate 2.4.1](https://github.com/whitesource/renovate-on-prem/releases/tag/2.4.1) works, v2.5.0 and v2.5.1 does not (so something between [32.185.3 and 32.105.0](https://octoclairvoyant.vercel.app/comparator?repo=renovatebot%2Frenovate&to=32.185.3&from=32.105.0)).

## Logs

Whitesource Renovate 2.4.1 (working):

```
"msg":"No node constraint found - using latest"
"msg":"Using buildpack dynamic installs"
"msg":"Resolved stable matching version", "toolName":"npm", "resolvedVersion":"8.19.2"
"msg":"Executing command", "command":"install-tool npm 8.19.2"
"msg":"LOG_CONTEXT: undefined"
"msg":"Ignoring non-installed repo"
"msg":"exec completed", "cmd":"install-tool npm 8.19.2", "durationMs":8092,"stdout":"installing v2 tool npm v8.19.2\n\nadded 1 package in 7s\nlinking tool npm v8.19.2\n8.19.2\n", "stderr":""
"msg":"Executing command", "command":"hash -d npm 2>/dev/null || true"
"msg":"exec completed", "cmd":"hash -d npm 2>/dev/null || true", "durationMs":74, "stdout":"", "stderr":""
```

Whitesource Renovate 2.5.1 (failing):

```
"msg":"No node constraint found - using latest"
"msg":"Setting BUILDPACK_CACHE_DIR to /tmp/renovate/containerbase"
"msg":"Using containerbase dynamic installs"
"msg":"Host is disabled - rejecting request", "url":"https://nodejs.org/dist/index.json"
"err":{"message":"No tool releases found.","stack":"Error: No tool releases found.\n at resolveConstraint (/usr/src/app/node_modules/renovate/dist/util/exec/containerbase.js:166:15)\n at async generateInstallCommands (/usr/src/app/node_modules/renovate/dist/ut
"msg":"No updated lock files in branch"
```

So it seems that something has been changed to how Node is resolved (may it be the [change of Node versioning](https://github.com/renovatebot/renovate/pull/17014)?), and that we now fall short of the lack of custom registry support for Node datasource (like in https://github.com/renovatebot/renovate/discussions/17518)?

## Workaround in v2.5.0/v2.5.1

If I add a Node constraint to only use the Node version that is in v2.5.0/v2.5.1, it works, in `package.json`:

```json
  "engines": {
    "node": "16.17.0"
  },
```

## Question / discussion

Is this a bug / regression, or intended behaviour?

**Comment by rarkins** [comment]
Thank you for raising a discussion first. Although failures are never intended, essentially this is intended _behavior_, which I will explain:

- Renovate needs to run CLI tools like npm, gradle, etc
- Most self-hosted users do not have a single version of each tool in use, e.g. might have node 14 and node 16, etc
- Renovate is well progressed towards an approach of "dynamically" installing all tools on-demand, e.g. the exact version required by a project. Not all tools are supported yet, but Node and npm are two which are
- If a project does not have an exact version of a tool or language specified, Renovate will fetch a list of versions for that tool to decide which to use, e.g. latest or ^16.17.0, and then install the one it decides
- In the Node.js case, it's using the Node.js datasource which is sourced from nodejs.org
- You have blocked the nodejs.org server, so it cannot fetch the list of versions and fails

By specifying an exact version, Renovate did not need to fetch the list of versions from nodejs.org, so it worked around the problem. Possibly if you specified a *different* version of Node it may have failed when it tried to download it - I'm not sure though as it's environmental.If so then another workaround may be to set `RENOVATE_BINARY_SOURCE=global` in env so that Renovate never dynamically installs tools and uses the default ones installed within the image. In this case you're stuck with the version of tools in the image each release.

**Reply by rarkins** [reply]
Further, we are open to trying to adjust these processes in restricted environments if it can be done elegantly, e.g. alternative datasources, download locations, etc. We don't have an anti-airgap philosophy but it's also challenging when we *need* to download things to meet certain requirements.

**Reply by mikaello** [reply]
Thanks a lot for the swift and thorough explanation!

> Possibly if you specified a different version of Node it may have failed when it tried to download it

I can confirm that, even specifying something like `16.x` failed.

So the reason it worked before v2.5.1 was that the Node version was not correctly resolved, and then used the default version?

I can confirm that setting `RENOVATE_BINARY_SOURCE=global` works in v2.5.1, but from what I read from [the documentation](https://docs.renovatebot.com/self-hosted-configuration/#binarysource), that will affect all kind of third party tools, right?

The possibility of providing a custom registry for these CLI tools would be very nice. In Artifactory it is possible to create remotes that proxy a custom data source, like https://nodejs.org/dist.

**Reply by mikaello** [reply]
Reply to self:

> I can confirm that setting RENOVATE_BINARY_SOURCE=global works in v2.5.1, but from what I read from [the documentation](https://docs.renovatebot.com/self-hosted-configuration/#binarysource), that will affect all kind of third party tools, right?

Yes, it affects all the bundled CLI tools. After some more testing, this actually solves a few more problems we have experienced, suddenly Poetry updates started working (before `https://api.github.com/graphql` was queried, but that is of course blocked in the air gapped environment). 👍 

Unfortunately, `RENOVATE_BINARY_SOURCE=docker` is not an option in the runtime environment of our Renovate instance.

**Reply by rarkins** [reply]
Do you mean that you can essentially proxy any http datasource? e.g. instead of `https://nodejs.org/dist` you might have `https://artifactory.company.com/proxy/nodejs/dist/` ?

**Reply by mikaello** [reply]
Yes, it works by creating a "[remote repository](https://www.jfrog.com/confluence/display/JFROG/Remote+Repositories)" in Artifactory for each datasource that you be possible to query.

So typically you will create remote repositories for common APIs and package repositories, like `api.github.com` -> `artifactory.company.com/github-remote` (and similar for e.g. Maven repositories, but then it will not act as (more or less) pure proxy, but add more functionality). For `https://nodejs.org/dist` a remote repository as you exemplified can be created.

**Reply by rarkins** [reply]
Despite the fancy name "remote repository", it just supports plain HTTP forwarding?

Maybe a lower level solution for us would be a new hostRules entry like `replaceHost`. e.g.:

```json
{
  "matchHost": "nodejs.org",
  "replaceHost": "https://artifactory.company.com/proxy/nodejs"
}
```

**Reply by mikaello** [reply]
Such a low level solution would indeed solve the problem in this case, and in other cases (e.g. obsolete the need for Git smudge filters to alter URLs (ref. [Git smudge filter discussion](https://github.com/renovatebot/renovate/discussions/17932))).

> Despite the fancy name "remote repository", it just supports plain HTTP forwarding?

As a user (like Renovatebot is), and for all practical concerns in this context, that is correct.

**Reply by rarkins** [reply]
In this case feel free to raise a feature request (and/or PR) for the replaceHost idea. Hopefully it's as simple as I expect

**Reply by mikaello** [reply]
Sorry for the slow response. I was just about to create a feature request, but found that this very feature request is already proposed in this issue: https://github.com/renovatebot/renovate/issues/12042#issuecomment-969970598

**Reply by rarkins** [reply]
Totally forgot about that thread!
